### PR TITLE
Feature/iterator variables

### LINKED ISSUE #7: Iterator variables
Iterator variables are not taken into account for the moment for the semantic analysis

### PR DESCRIPTION
**Pull request by aitorres**
Fixes #7, adds semantic checking to iteration variables

### CODE REVIEW

File: src/Parser.y [hunk 1]
```diff
@@ -608,6 +610,22 @@ checkConstantReassignment e = case G.expAst e of
...
+  else do
+    let (G.InstAsig f _) = head errors
+    let faultyTk = G.expTok f
+    RWS.tell [ST.SemanticError ("Iteration variable " ++ show id ++ " must not be reassigned") faultyTk]
+    return ()
+  where isAssignment (G.InstAsig _ _) = True
+        isAssignment _ = False
+        modifiesIterVariable (G.InstAsig l _) = (T.cleanedString (G.expTok l)) == (T.cleanedString tk)
+        modifiesIterVariable _ = False
+
```
Comment (german1608) [inline src/Parser.y]: This works but it's not the best solution. The main problem is that iterator variables are not being inserted to symtable. We need to do so and mark the variable as a constant or something, so when we parse the CODEBLOCK of the for structure it checks that the variable reassigned is indeed a iterator variable and raise an error if so.

**Review by german1608 — CHANGES_REQUESTED** [review]

File: src/Parser.y (hunk 1 above)
Comment (german1608) [reply src/Parser.y]: with this program:
```
hello ashen one

traveling somewhere

   upgrading i with 1 soul until level 20
      traveling somewhere
         with orange saponite say i \
         with orange saponite say |\n|
      you died
   max level reached

you died

farewell ashen one
```

the program fails on the 7th line, pointing to the `i` variable usage though its a valid program

Comment (german1608) [reply src/Parser.y]: with this program:
```
hello ashen one

traveling somewhere

   upgrading i with 1 soul until level 20
      traveling somewhere
         with orange saponite say i \
         with orange saponite say |\n|
      you died
   max level reached

you died

farewell ashen one
```

the program fails on the 7th line, pointing to the `i` variable usage though its a valid program

Comment (german1608) [reply src/Parser.y]: image

**Comment by aitorres** [comment]
After revising our language specs and the way we handle iterations, both bounded and structured iterations, we have compromised on a solution to properly keep track of iteration variables.

File: src/Parser.y [hunk 2]
```diff
@@ -418,8 +418,40 @@ INSTR :: { G.Instruction }
...
+FORBEGIN :: { G.Id }
+  : forBegin ID                                                         {% do
+                                                                          mid <- checkIdAvailability $2
+                                                                          (ST.SymTable d s cs ivs) <- RWS.get
+                                                                          case mid of
+                                                                            Just ST.DictionaryEntry {ST.name=varName} -> do
+                                                                              RWS.put (ST.SymTable d s cs (varName:ivs))
+                                                                            Nothing -> do
+                                                                              -- checkIdAvailability already rose an error here, nothing else to do
+                                                                              RWS.put (ST.SymTable d s cs ("_UNDECLARED_ITER_VAR":ivs))
```
Comment (german1608) [inline src/Parser.y]: why do we add `"_UNDECLARED_ITER_VAR"` here?

File: src/Parser.y [hunk 3]
```diff
@@ -418,8 +418,40 @@ INSTR :: { G.Instruction }
...
+FOREACHBEGIN :: { G.Id }
+  : forEachBegin ID                                                     {% do
+                                                                          mid <- checkIdAvailability $2
+                                                                          (ST.SymTable d s cs ivs) <- RWS.get
+                                                                          case mid of
+                                                                            Just ST.DictionaryEntry {ST.name=varName} -> do
+                                                                              RWS.put (ST.SymTable d s cs (varName:ivs))
+                                                                            Nothing -> do
+                                                                              -- checkIdAvailability already rose an error here, nothing else to do
+                                                                              RWS.put (ST.SymTable d s cs ("_UNDECLARED_STRUCTITER_VAR":ivs))
```
Comment (german1608) [inline src/Parser.y]: Same comment here

File: src/Parser.y [hunk 4]
```diff
@@ -543,7 +575,9 @@ addIdToSymTable mi d@(c, gId@(G.Id tk@(T.Token {T.aToken=at, T.cleanedString=idN
       else if category == ST.Function
       then RWS.tell $ [ST.SemanticError ("Name " ++ show tk ++ " conflicts with a function") tk]
       else if category == ST.RefParam || category == ST.ValueParam
-      then RWS.tell $ [ST.SemanticError ("Name "++ show tk ++ " conflicts with a formal param") tk]
+      then RWS.tell $ [ST.SemanticError ("Name " ++ show tk ++ " conflicts with a formal param") tk]
+      else if elem (T.cleanedString tk) iterVars
```
Comment (german1608) [inline src/Parser.y]: better to use `elem` as an "operator"
```haskell
`elem`
```

**Review by german1608 — CHANGES_REQUESTED** [review]

File: src/Parser.y (hunk 2 above)
Comment (aitorres) [reply src/Parser.y]: removed

File: src/Parser.y (hunk 3 above)
Comment (aitorres) [reply src/Parser.y]: removed

File: src/Parser.y (hunk 4 above)
Comment (aitorres) [reply src/Parser.y]: changed in both uses that arised from this PR

**Comment by aitorres** [comment]
Screenshot_20200114_213358

Constants are now being checked also.

**Review by german1608 — APPROVED** [review]

**Pull request merged**